File: src/Chapter05GraphModellingAndFloodFill/Leetcode130.java
package Chapter05GraphModellingAndFloodFill;

/**
 * Author : lihuichuan
 * Time   : 2019/12/22
 */

/// Floodfill
/// Time Complexity: O(n * m)
/// Space Complexity: O(n * m)
public class Leetcode130 {
    private int d[][] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    private int m, n;
    private boolean visited[][];

    public void solve(char[][] board) {
        if (board == null || board.length == 0 || board[0].length == 0)
            return ;
        m = board.length;
        n = board[0].length;

        visited = new boolean[m][n];
        for(int i = 0; i < m; i ++){
            if(board[i][0] == 'O') dfs(board, i, 0);
            if(board[i][n - 1] == 'O') dfs(board, i, n - 1);
        }
        for(int j = 0; j < n; j ++){
            if(board[0][j] == 'O') dfs(board, 0, j);
            if(board[m - 1][j] == 'O') dfs(board, m - 1, j);
        }
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
               if(board[i][j] == 'O'){
                   board[i][j] = 'X';
               }else if(board[i][j] == '+'){
                   board[i][j] = 'O';
               }
    }

    public void dfs(char[][] grid, int i, int j) {

        grid[i][j] = '+';
        visited[i][j] = true;
        for (int index = 0; index < d.length; index++) {
            int newx = i + d[index][0];
            int newy = j + d[index][1];
            if (inArea(newx, newy) && !visited[newx][newy] && grid[newx][newy] == 'O') {
                dfs(grid, newx, newy);
            }
        }
        return;
    }

    private boolean inArea(int x, int y) {
        return x >= 0 && x < m && y >= 0 && y < n;
    }
}

File: src/Chapter05GraphModellingAndFloodFill/python/leetcode130.py
class Solution:
    postion = [[0, 1], [1, 0], [0, -1], [-1, 0]]

    def solve(self, board):
        if not board:
            return board
        visited = [[False for _ in range(len(board[i]))] for i in range(len(board))]
        res = []
        m = len(board)
        n = len(board[0])
        for j in range(n):
            self._dfs(board, 0, j, visited)
            self._dfs(board, m - 1, j, visited)

        for i in range(m):
            self._dfs(board, i, 0, visited)
            self._dfs(board, i, n - 1, visited)

        for i in range(m):
            for j in range(n):
                if board[i][j] == "O":
                    board[i][j] = "X"
                elif board[i][j] == "+":
                    board[i][j] = "O"

    def _dfs(self, board, i, j, visited):
        visited[i][j] = True
        if board[i][j] == 'O':
            board[i][j] = '+'

            for next_step in self.postion:
                next_x = i + next_step[0]
                next_y = j + next_step[1]
                if next_x >= 0 and next_x < len(board) and next_y >= 0 and next_y < len(board[0]) and not \
                        visited[next_x][next_y]:
                    self._dfs(board,next_x, next_y, visited)